src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.h

```diff
@@ -0,0 +1,219 @@
+     * @param shadow_distance The distance our friendly robot will position itself away
+     * from the shadowee
+     */
+    static Point findBlockShotPoint(const Robot &robot, const Field &field,
+                                    const Team &friendlyTeam, const Team &enemyTeam,
+                                    const Robot &shadowee, const double &shadow_distance)
+    {
+        std::vector<Robot> robots_to_ignore = {robot};
+        if (friendlyTeam.goalie().has_value())
+        {
+            robots_to_ignore.emplace_back(friendlyTeam.goalie().value());
+        }
+
+        auto best_enemy_shot_opt =
+            calcBestShotOnGoal(field, friendlyTeam, enemyTeam, shadowee.position(),
+                               TeamType::FRIENDLY, robots_to_ignore);
+
+        Vector enemy_shot_vector = Vector(0, 0);
+        if (best_enemy_shot_opt)
+        {
+            enemy_shot_vector =
+                best_enemy_shot_opt->getPointToShootAt() - shadowee.position();
+        }
+        else
+        {
+            enemy_shot_vector = field.friendlyGoalCenter() - shadowee.position();
+        }
+        return shadowee.position() + enemy_shot_vector.normalize(shadow_distance);
+    }
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto block_shot_s     = state<BlockShotState>;
+        const auto block_pass_s     = state<BlockPassState>;
+        const auto steal_and_chip_s = state<StealAndChipState>;
+
+        const auto update_e = event<Update>;
+
+        /**
+         * Guard that checks if the enemy threat has ball
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_has_ball = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return enemy_threat_opt.value().has_ball;
+            };
+            return false;
+        };
+
+        /**
+         * Guard that checks if the robot is done moving
+         * to the block shot point position calculated in
+         * findBlockShotPoint()
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the robot's current position matches the one
+         *         returned from findBlockShotPoint
+         */
+        const auto is_blocking_shot = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                // We compare the length between the two points rather than using the
+                // equality operator for points as the robot does not end up in the exact
+                // position as findBlockShotPoint
+                return Segment(event.common.robot.position(),
```

Comment:
I'm pretty confused by this Segment. Can't we subtract one point from the other and check the length?

Reply:
Removed entire function since I transitioned to MoveFSM